presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java

```diff
@@ -0,0 +1,393 @@
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+
+public class PushDownDereferences
+{
+    private final Metadata metadata;
+    private final SqlParser sqlParser;
+
+    public PushDownDereferences(Metadata metadata, SqlParser sqlParser)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+    }
+
+    public Set<Rule<?>> rules()
+    {
+        return ImmutableSet.of(
+                new DereferencePushDownJoin(metadata, sqlParser),
+                new DereferencePushDownSort(metadata, sqlParser),
+                new DereferencePushDownProject(metadata, sqlParser));
+    }
+
+    public abstract class DereferencePushDownRule<N extends PlanNode>
+            implements Rule<ProjectNode>
+    {
+        private final Capture<N> targetCapture = newCapture();
+        private final Pattern<N> targetPattern;
+
+        protected final Metadata metadata;
+        protected final SqlParser sqlParser;
+
+        protected DereferencePushDownRule(Metadata metadata, SqlParser sqlParser, Pattern<N> targetPattern)
+        {
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+            this.targetPattern = requireNonNull(targetPattern, "targetPattern is null");
+        }
+
+        @Override
+        public Pattern<ProjectNode> getPattern()
+        {
+            return project().with(source().matching(targetPattern.capturedAs(targetCapture)));
+        }
+
+        @Override
+        public Result apply(ProjectNode node, Captures captures, Context context)
+        {
+            N child = captures.get(targetCapture);
+            Map<DereferenceExpression, Symbol> expressions = getDereferenceSymbolMap(node, context, metadata, sqlParser);
+            Assignments assignments = node.getAssignments().rewrite(new DereferenceReplacer(expressions));
+
+            Result result = pushDownDereferences(context, child, expressions, assignments);
+            if (result.isEmpty()) {
+                return Result.empty();
+            }
+            return Result.ofPlanNode(new ProjectNode(context.getIdAllocator().getNextId(), result.getTransformedPlan().get(), assignments));
+        }
+
+        protected abstract Result pushDownDereferences(Context context, N targetNode, Map<DereferenceExpression, Symbol> expressions, Assignments assignments);
+
+        protected boolean isCastRowType(DereferenceExpression dereference)
+        {
+            Expression base = dereference.getBase();
+            if (base instanceof Cast) {
+                return metadata.getTypeManager().getType(parseTypeSignature(((Cast) base).getType())) instanceof RowType;
+            }
+            return false;
+        }
+    }
+
+/**
+ * Transforms:
+ * <pre>
+ *  Project(a.msg.x)
+ *    Join(a.msg.y = b.msg.y)
```

Comment:
This is confusing because join's equi-clause can't contain expressions. The following might be clearer:

```
     * Transforms:
     * <pre>
     *  Project(a_x := a.msg.x)
     *    Join(a_y = b_y) => [a]
     *      Project(a_y := a.msg.y)
     *          Source(a)
     *      Project(b_y := b.msg.y)
     *          Source(b)
     *  </pre>
     * to:
     * <pre>
     *  Join(a_y = b_y) => [a_x]
     *    Project(a_x := a.msg.x, a_y := a.msg.y)
     *      Source(a)
     *    Project(b_y := b.msg.y)
     *      Source(b)
     * </pre>
```